### PR TITLE
Add support for Clang-Tidy and additional Clang warnings

### PR DESCRIPTION
**Pull request by agarny**
Addresses issue #283.

Clang-Tidy and the additional Clang warnings are used for our `src` and `tests` directories. However, when it comes to the `tests` directory, we had to consider `gtest/` as a system header prefix, to prevent Clang from generating an awful lot of warnings. In the same way, we had to disable some warnings in Clang-Tidy, so that we can "safely" use  `TEST()` and `EXPECT_EQ()`. Also, note that in order to use Clang-Tidy, we had to bump our minimum requirement for CMake. So, now, we need at least CMake 3.6 (while 3.5.2 before).

Note that Clang-Tidy and the additional Clang warnings only make sense on our macOS build machine. Actually, Clang-Tidy could be installed on our Windows and Linux build machines too, if we wanted, but Clang is not used on those machines anyway, so probably best to just stick to our macOS build machine to get a full report of what could be improved in our codebase, if anything, not least since our macOS build machine is the fastest of all three.

The bottom line is that we are now checking an awful lot of things, which should make our codebase pretty solid.

Otherwise, I have taken that opportunity to also:
 - clean up various bits of codes here and there;
 - move the search for LibXml2 from `CMakeLists.txt` to `cmake/common.cmake`, which I think is where it really ought to be?
 - upgrade our version of GTest. We are now using version 1.8.1. This wasn't strictly required, but at some point I thought upgrading GTest might prevent the thousands of warnings I was getting, but it didn't make any difference. Yet, I had upgraded, so we might as well use the latest version of GTest.

### CODE REVIEW

**Comment by agarny** [comment]
Argh, our Linux build machine only has CMake 3.5.2 installed on it. @hsorby, I thought it was bioeng49, but it clearly isn't since bioeng49 has CMake 3.12.4. So, which machine is used for building libCellML on Linux? In fact, could you upgrade its copy of CMake?

Regarding our macOS build machine, I also thought that it was bn337051, but once again it doesn't seem to be the case since it seems to be running macOS 10.12...? I installed Clang-Tidy on bn337051, but it looks like it might have been unnecessary. Still, for this PR to make any sense, we need Clang-Tidy installed on the macOS build machine used to build libCellML. Again, @hsorby, it would be nice to have that done.

In summary:
 - [ ] Have our Linux build machine use at least CMake 3.6.
 - [ ] Install Clang-Tidy on our macOS build machine (using `brew install llvm` and then symlinking `clang-tidy` (and `clang-format`) to `/usr/local/bin`).
 - [ ] Repeat commit 89cbd45 to make sure that our macOS build does indeed fail.
 - [ ] Make Clang and Clang-Tidy happy by committing something like 029a7e4.

**Comment by agarny** [comment]
Would be nice to have something done about this PR... especially with Keri now working on libCellML. I would also like to have `clang-format` used to ensure that our code is consistent throughout (see issue #284).

**Comment by nickerso** [comment]
Just to be clear, we're waiting on some upgrades to the build machines to get things playing nice with these changes, right?

**Comment by hsorby** [comment]
I believe you are right.

**Comment by agarny** [comment]
That is correct. The version of CMake on the Linux slave is too old for this PR to pass, and on the macOS slave we should have `clang-tidy` installed so that this PR can actually be useful.

I thought that those slaves were `bioeng49` and `bn337051`, but they are not. `bioeng49` runs Ubuntu 16.04 LTS while the Linux slave used by libCellML runs Ubuntu 14.04 LTS (?!). As for the macOS slave used by libCellML, it seems to be running macOS Sierra (10.12) while `bn337051` runs macOS High Sierra (10.13).

So, in the end, I have no idea what/where those slaves are... and therefore can't update them...

**Comment by hsorby** [comment]
The build slaves have now been updated and all pertinent builds have been triggered.  Failures now indicate true failures that need to be fixed.  This (I'm fairly confident) is the only PR needing changes.

**Comment by agarny** [comment]
Hmm, not quite sure how I missed that one, but I can indeed confirm the same build error on my local Ubuntu 18.04 LTS box. Ok, consider it fixed... soon...

**Comment by nickerso** [comment]
This one is good to merge in now, right? is it going to cause major headaches in the other open pull requests (especially #304) ?

**Comment by agarny** [comment]
> This one is good to merge in now, right? is it going to cause major headaches in the other open pull requests (especially #304) ?

As far as I am concerned, it is indeed ready to merge. And, yes, it's likely to cause some headache with all pending PRs...

**Comment by hsorby** [comment]
First in first served?  That is first to get their PR approved and merged passes on the burden of conflict resolution to PRs not yet approved.

**Review by nickerso — APPROVED** [review]

File: cmake/environmentchecks.cmake [hunk 1]
```diff
@@ -14,21 +14,54 @@
...
+  # we wish to use Config mode in the first call.  Finding LibXml2 in config mode
+  # is the preferred method so we will try this first quietly.
+  #
+  # This does change how we get information about include paths and such so we
+  # need to track how we found LibXml2.
+  find_package(LibXml2 CONFIG QUIET)
+  if (LibXml2_FOUND)
+    set(HAVE_LIBXML2_CONFIG TRUE)
+  else ()
+    find_package(LibXml2 REQUIRED)
```
Comment (hsorby) [inline cmake/environmentchecks.cmake:49]: Can we add in here:

`unset(LibXml2_DIR CACHE)`

Comment (agarny) [reply cmake/environmentchecks.cmake:49]: There are several calls to `find_package(LibXml2...)`. So, are you meaning that we should that `unset()` call between `set(HAVE_LIBXML2_CONFIG FALSE)`and `if(MSVC)`?

**Comment by agarny** [comment]
@hsorby, would you know why the Python tests fail on the macOS slave? FWIW, those tests work on my Mac...

**Comment by hsorby** [comment]
Yes, the macOS slave is running SWIG 4.0.  So we need to look at that.

**Comment by hsorby** [comment]
The new SWIG 4.0 has changed the way it deals with rpath I think so we need to upgrade the handling of the bound dynamic libraries.

**Comment by hsorby** [comment]
I think this one is good to merge now.

**Comment by agarny** [comment]
FWIW, I want to go through this one more time (today). There have been quite a few `git merge`s lately. So, I want to make sure that it all looks good.

**Comment by agarny** [comment]
Actually, @hsorby, don't we want to create a test for Clang-Tidy in the same that you created one for ClangFormat?

**Comment by hsorby** [comment]
The problem with the clang-tidy is that it is a compiler level change.  It is not really external.  The clang tidy then becomes part of the standard warnings and errors that we already deal with.  So I think we just add it in when available the Buildbot testing will do the same and our checks should continue as they currently are with no further changes necessary.

**Comment by agarny** [comment]
> The problem with the clang-tidy is that it is a compiler level change. It is not really external. The clang tidy then becomes part of the standard warnings and errors that we already deal with. So I think we just add it in when available the Buildbot testing will do the same and our checks should continue as they currently are with no further changes necessary.

Ok, fair enough.

Otherwise, I have just gone through this PR one more time and I am happy with it. So, up to the reviewers now... :)

**Comment by agarny** [comment]
From what I can tell, both @hsorby and @nickerso are OK with this PR. So, either of you can merge it in...
